language/bytecode-verifier/src/abstract_state.rs

```diff
@@ -105,489 +133,277 @@ impl AbstractState {
-
-        AbstractState {
-            locals,
-            globals,
-            borrows,
-            partition,
-        }
+        nonce_map.insert(self.frame_root, self.frame_root);
+        let canonical_state = AbstractState {
+            locals: new_locals,
+            borrow_graph: self.borrow_graph.rename_nonces(nonce_map),
+            frame_root: self.frame_root,
+            next_nonce: self.frame_root.inner() + 1,
+        };
+        checked_postcondition!(canonical_state.is_canonical());
+        canonical_state
     }
 
-    fn unrestricted_borrowed_value_unavailable(
-        state1: &AbstractState,
-        state2: &AbstractState,
-    ) -> bool {
-        state1.locals.keys().any(|idx| {
-            state1.locals[idx].is_unrestricted_value()
-                && state1.locals[idx].is_borrowed_value()
-                && !state2.locals.contains_key(idx)
+    fn all_nonces_immutable(&self, borrows: BTreeSet<Nonce>) -> bool {
+        !self.locals.values().any(|abs_type| {
+            abs_type.signature.is_mutable_reference()
+                && borrows.contains(&abs_type.value.extract_nonce().unwrap())
         })
     }
 
-    fn strong_propagate(
-        &self,
-        nonce: Nonce,
-        borrow_info: &BorrowInfo,
-    ) -> BTreeMap<Nonce, BorrowInfo> {
-        let mut new_borrows = BTreeMap::new();
-        let mut singleton_nonce_set = BTreeSet::new();
-        singleton_nonce_set.insert(nonce);
-        for (src_nonce, borrowed_by) in &self.borrows {
-            if self.partition.is_equal(*src_nonce, nonce) {
-                if let BorrowInfo::BorrowedBy(nonce_set) = borrowed_by {
-                    if nonce_set == &singleton_nonce_set {
-                        new_borrows.insert(*src_nonce, borrow_info.clone());
-                    }
-                }
-            }
-        }
-        new_borrows
-    }
-
-    fn borrow_from_nonce(&mut self, nonce: Nonce, new_nonce: Nonce) {
-        self.borrows.entry(nonce).or_insert({
-            let mut nonce_set = BTreeSet::new();
-            nonce_set.insert(new_nonce);
-            BorrowInfo::BorrowedBy(nonce_set)
-        });
+    fn is_canonical(&self) -> bool {
+        self.locals.iter().all(|(x, y)| {
+            !y.value.is_reference() || Nonce::new(*x as usize) == y.value.extract_nonce().unwrap()
+        })
     }
 
-    fn map_nonce_set(
-        nonce_map: &BTreeMap<Nonce, Nonce>,
-        nonce_set: &BTreeSet<Nonce>,
-    ) -> BTreeSet<Nonce> {
-        let mut mapped_nonce_set = BTreeSet::new();
-        for nonce in nonce_set {
-            mapped_nonce_set.insert(nonce_map[nonce]);
-        }
-        mapped_nonce_set
+    fn borrowed_value_unavailable(state1: &AbstractState, state2: &AbstractState) -> bool {
+        state1.locals.keys().any(|idx| {
+            state1.locals[idx].value.is_value()
+                && state1.is_local_borrowed(*idx)
+                && !state2.locals.contains_key(idx)
```

Comment:
This is a good case. I did not think clearly about this case. 
I think this case means you cannot do locals and borrows in a separate pass. 
(Typing can still be done in a separate pass, but locals and borrows are entangled because of this case

Reply:
This is a great case. 
It also means that JoinErrors are a lot harder to get rid of